src/network/network_crypto.cpp

```diff
@@ -0,0 +1,112 @@
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file network_admin.cpp Server part of the admin network protocol. */
+
+#include "../stdafx.h"
+
+#include "network_crypto.h"
+#include "core/packet.h"
+#include "../core/random_func.hpp"
+#include "../3rdparty/monocypher/monocypher.h"
+#include "../settings_type.h"
+#include "../string_func.h"
+
+#include "../safeguards.h"
+
+/**
+ * Call \c crypto_wipe for all the data in the given span.
+ * @param span The span to cryptographically wipe.
+ */
+static void crypto_wipe(std::span<uint8_t> span)
+{
+	crypto_wipe(span.data(), span.size());
+}
+
+static ssize_t TransferOutSpan(std::span<uint8_t> destination, const char *source, size_t amount)
+{
+	size_t to_copy = std::min(amount, destination.size());
+	std::copy(source, source + to_copy, destination.data());
+	return to_copy;
+}
+
+template <size_t size>
+class SharedKeys {
+	std::array<uint8_t, 2 * size> keys;
+public:
+	~SharedKeys() { crypto_wipe(keys); }
+	constexpr std::span<const uint8_t> ClientToServer() const { return std::span(this->keys.data(), size); }
+	constexpr std::span<const uint8_t> ServerToClient() const { return std::span(this->keys.data() + size, size); }
+	constexpr std::span<uint8_t> Both() { return this->keys; }
+};
+
+X25519Key::~X25519Key()
+{
+	crypto_wipe(*this);
+}
+
+static bool X25519KeyExchange(SharedKeys<32> &shared_keys, const X25519Key &my_private_key, const X25519Key &their_public_key)
+{
+	X25519Key shared_secret;
+	crypto_x25519(shared_secret.data(), my_private_key.data(), their_public_key.data());
+	if (std::all_of(shared_secret.begin(), shared_secret.end(), [](auto v) { return v == 0; })) {
+		/* Secret is all 0 because public key is all 0, just reject it */
+		return false;
+	}
+
+	crypto_blake2b_ctx ctx;
+	crypto_blake2b_init(&ctx, shared_keys.Both().size());
+	crypto_blake2b_update(&ctx, shared_secret.data(), shared_secret.size());
+	crypto_blake2b_update(&ctx, my_public_key.data(), my_public_key.size());
+	crypto_blake2b_update(&ctx, their_public_key.data(), their_public_key.size());
+	crypto_blake2b_final(&ctx, shared_keys.Both().data());
+	return true;
+}
+
+
+X25519AuthorizedKeyHandler::X25519AuthorizedKeyHandler()
+{
+	RandomBytesWithFallback(this->my_private_key);
+	crypto_x25519_public_key(this->my_public_key.data(), this->my_private_key.data());
+
+	RandomBytesWithFallback(this->nonce);
+}
+
+X25519AuthorizedKeyHandler::~X25519AuthorizedKeyHandler()
+{
+	crypto_wipe(this->nonce);
+}
+
+/* virtual */ void X25519AuthorizedKeyHandler::SendRequest(Packet &p)
+{
+	p.Send_bytes(this->my_public_key);
+	p.Send_bytes(this->nonce);
```

Comment:
From a lot of examples I have seen, normally the nonce is send with the mac. I don't think it hurts to do it this way; but it is a bit unusual.

The main reason I can imagine this being better, is that we now control the nonce; otherwise the client might be reusing theirs, which is an absolute no-no.
So I can imagine you did it for that reason. Just checking in, basically.

Crypto is hard. My head hurts. :P

Reply:
Yes, I sent the nonce here to ensure it to have at least some random-ness, instead of hoping the authors of the ottd-admin libraries are going to do the right thing.